osu.Game/Screens/Edit/HitObjectChangeHandler.cs

```diff
@@ -0,0 +1,293 @@
+        /// </summary>
+        private readonly EditorChangeHandler? changeHandler;
+
+        public readonly Bindable<bool> CanUndo = new BindableBool();
+
+        public readonly Bindable<bool> CanRedo = new BindableBool();
+
+        public event Action? OnStateChange;
+
+        public bool HasUncommittedChanges => currentTransaction.UndoChanges.Count != 0;
+
+        public Guid CurrentState => undoStack.TryPeek(out var transaction) ? transaction.Id : Guid.Empty;
+
+        private Transaction currentTransaction;
+
+        private readonly Stack<Transaction> undoStack = new Stack<Transaction>();
+
+        private readonly Stack<Transaction> redoStack = new Stack<Transaction>();
+
+        private bool isRestoring;
+
+        public HitObjectChangeHandler(EditorBeatmap editorBeatmap, EditorChangeHandler? changeHandler = null)
+        {
+            currentTransaction = new Transaction();
+            this.editorBeatmap = editorBeatmap;
+            this.changeHandler = changeHandler;
+
+            editorBeatmap.TransactionBegan += BeginChange;
+            editorBeatmap.TransactionEnded += EndChange;
+            editorBeatmap.SaveStateTriggered += SaveState;
+
+            if (this.changeHandler != null)
+                this.changeHandler.OnStateChange += commitChangeHandlerStateChange;
+        }
+
+        /// <summary>
+        /// Submits a change to be applied and added to the history.
+        /// </summary>
+        /// <param name="change">Change to be applied.</param>
+        /// <param name="commitImmediately">Whether to commit the current transaction and push it onto the undo stack immediately.</param>
+        public void Submit(IRevertibleChange change, bool commitImmediately = false)
+        {
+            if (commitImmediately)
+                BeginChange();
+
+            change.Apply();
+            Record(change);
+
+            if (commitImmediately)
+                EndChange();
+        }
+
+        /// <summary>
+        /// Submits a collection of changes to be applied and added to the history.
+        /// </summary>
+        /// <param name="changes">Changes to be applied.</param>
+        /// <param name="commitImmediately">Whether to commit the current transaction and push it onto the undo stack immediately.</param>
+        public void Submit(IEnumerable<IRevertibleChange> changes, bool commitImmediately = false)
+        {
+            foreach (var change in changes)
+                Submit(change);
+
+            if (commitImmediately)
+                UpdateState();
+        }
+
+        protected override void UpdateState()
+        {
+            if (isRestoring || !HasUncommittedChanges)
+                return;
+
+            undoStack.Push(currentTransaction);
+            redoStack.Clear();
+
+            Logger.Log($"Added {currentTransaction.UndoChanges.Count} change(s) to undo stack");
+
+            currentTransaction = new Transaction();
```

Comment:
How does `currentTransaction` work, exactly?

Is there no explicit "start transaction" primitive? This looks like you're always in some implicitly-defined ambient transaction that gets committed via this `UpdateState()` call, which may or may not come from somewhere? (See previous comments where I'm confused as to why `Submit()` is only calling this in one of the overloads.)

Reply:
There is always this ambient current transaction that can accept Changes even when no transaction is currently in progress. The alternative is only accepting changes after a BeginChange and throwing an exception or ignoring for Changes that arrive outside a transaction.

I went with this direction because the old change handler also supports undoable changes outside of a Begin/EndChange.  It lets you make changes to the hitobjects and call UpdateState to make a new undo step. This method is often used in the metadata edit code.